content/post/postgres/extension-preloading.md

```diff
@@ -0,0 +1,170 @@
+long as your module is installed in the usual location via the [`MODULES`]
+`Makefile` variable. Still, it will be useful to document these options so
+that DBAs quickly see how to set things up for the users who need them.
+
+### Local Preloading
+
+As a special case, a DBA might want to make your debugging or
+performance-measurement extension available to any user who needs it, even
+unprivileged users. All they need to do is move it from `$libdir` to
+`$libdir/plugins`.
+
+Then any Postgres user can load it via the [`LOAD`] command or include it in
+their [`local_preload_libraries`] configuration via either [`PGOPTIONS`] or,
+for every connection, via `ALTER ROLE SET`:
+
+```sql
+ALTER ROLE role_name
+  SET local_preload_libraries TO '$libdir/plugins/mylib';
+```
+
+As an extension author, you don't need to do any special configuration; there
+is no `Makefile` variable to install it in the the `$libdir/plugins`
+directory. But it might be handy for DBAs to document this option *in addition
+to* the [session preloading](#session-preloading) options. But emphasize that
+it should be used if and *only if* they want to allow any and all of their
+users to load your extension library without any further intervention.
+
+### Shared Preloading
+
+The last preloading variable is [`shared_preload_libraries`], which should be
+required for modules to run in every session or to perform operations only
+available at service start up, such as [shared memory and lightweight locks]
+or starting [background workers].
+
+As an extension author, if your extension requires `shared_preload_libraries`
+preloading, the documentation should say so explicitly, and explain why. For
+examples of wording, see [pg_stat_statements], [sepgsql], and [auth_delay].
+
+Beyond these limited cases, any other modules can be added to
+[`shared_preload_libraries`] for efficiency purposes. However, since shared
+preload modules are loaded into every server process --- even if that process
+never uses the library --- preloading them is recommended only for libraries
+used in most sessions.
+
+As an extension author, it will be kind to DBAs to include this information,
+and to describe the circumstances under which they *might* want to preload
+your library in every service --- along with the caveat that doing so requires
+a server restart. For example wording, see [PL/Perl] and [auto_explain]
+
+### A Final Note
+
+One more requirement to carefully document is dependence on shared libraries
+provided by *other* extensions. For example, [Shaun Thomas] tells me that the
+[BDR] extension used to rely on [pglogical] being loaded first. The
+implication for preloading was that [pglogical] had to appear in the
+`shared_preload_libraries` variable *before* [BDR].
+
+So if your preload-requiring extension depends on other extensions, be sure to
+document the importance of load order and the impact on the format of the 
+`shared_preload_libraries` variable.
```

Comment:
This is more generally true, particularly when it comes to the definitions and execution order of hooks.  Since the order of the libraries in shared_preload_libraries impacts the order that the `_PG_init()` calls (where the hooks are normally set up), this directly impacts the hook chains.

The sort of cooperative hooks that Postgres has means that you count on the hook before you to call the hook itself, so depending on what happens in the hooks, later hooks may not even be called, which can cause issues down the road.

The workaround that some modules use is inspecting the `shared_preload_libraries` directive directly and throwing an error or warning if they are not in the expected load slot. (I want to say that `pgaudit` and `citus` do this, but am not 100% certain on that.)

Reply:
Well yeah, but that's really a case of an extension bug. In addition to not continuing the hook chain, there are tons of other ways extensions could break things, so I'm not sure it's useful to call this particular thing out. I think the lesson here is anytime you hook into Postgres itself, you should really know what you are doing and be extra careful.

Reply:
Oh, so if the extension is implemented properly, load order shouldn't matter?

Reply:
Okay, based on the [Slack discussion](https://postgresteam.slack.com/archives/C056ZA93H1A/p1720811692617319), in bc13ecc I tweaked this section, but mostly punted on any recommendations, since it will very much depend on the combination of libraries loaded and the order in which they call hooks. What a nightmare.